### ISSUE TITLE
Fix loss function

### ISSUE THREAD

**Issue by shouldsee** [IpsumDominum/Pytorch-Simple-Transformer#3]
The calculation of loss function is wicked and allows model to fit very quickly and produce very good translation. But unfortunately this is not real and the model is not able to spell out the German sentence itself. It could only completes an i-length German sentence if you give it the first (i-1) token, which means it could not generate the whole sentence from a start-of-sentence tag <sos>. 

It seems reasonable enough to use this loss to train the  network, but unreasonable to assess its translation ability, though I have yet to train this network to its full functionality.

```python


###################
## original leave-last-token out decoder, 
## Not sure what's the exact error in this calculation, 
## but maybe because the model see the mask token directly?
##
#Output German, One Token At A Time
all_outs = torch.tensor([],requires_grad=True).to(device)
for i in range(item["german"].shape[1]-1):
    out = model(item["german"][:,:i+1])
    all_outs = torch.cat((all_outs,out),dim=1)


# ###################
# My variation of leave-last-token-out decoder, Used at training
# output_vocab_size = german_vocab_len

g = item["german"].shape
x = torch.zeros( [g[0],g[1],],dtype=torch.long ).to(device)
all_outs = torch.tensor([],requires_grad=True).to(device)
for i in range(item["german"].shape[1]-1):
    xx = torch.zeros( [g[0],g[1], ],dtype=torch.long ).to(device)
    out = model(x)
    xx[:,i:i+1] = item["german"][:,i:i+1]
    x = x+xx
    all_outs = torch.cat((all_outs,out),dim=1)

# ###################
# My variation of beam search decoder
model.encode(item["english"][:,1:-1])
g = item["german"].shape
x = torch.zeros( [g[0],g[1],],dtype=torch.long ).to(device)
all_outs = torch.tensor([],requires_grad=True).to(device)
for i in range(item["german"].shape[1]-1):
    out = model(x)
    x[:,i:i+1] = out.argmax(axis=-1)
    all_outs = torch.cat((all_outs,out),dim=1)

```

I found this glitch when fiddling with the attention layer at its core, and found zeroing the attention value created no harm to the performance of a last-token-only model in 

`sub_layers.py`

```python
attention_weights = F.softmax(attention_weights,dim=2)
attention_weights = attention_weights *0.   ## Try this!
```

**Comment by IpsumDominum** [comment]
Hi there. Thanks for the contribution again :D

The model shouldn't see any masked tokens directly, the "value" item of the first multi headed attention block encountered in the decoder gets only the last item plucked from the previous output. See code below. Unless you are talking about something else?

```python3
 def forward(self, encoder_outs, x):
        for block in self.transformer_blocks:
            output_seq_attention_out = self.multi_head_attention(
                query=x[:, -1:, :], key=x, value=x, residual_x=x[:, -1:, :]
            )
            x = block(
                query=output_seq_attention_out,
                key=encoder_outs,
                value=encoder_outs,
                residual_x=output_seq_attention_out,
            )
        return self.vocab_logits(x)
```

The query and key matmul in the decoder should have the triangular mask applied such that the timestep i term in the value matrix won't be able to access information of query * key pairs which correspondingly have timesteps before i. I'm using timestep for the index in the sequence num. 

Specifically, consider the output matrix from query matmul key = 
[ q1 * k1, q1 * k2, q1 * k3...
  q2 * k1, q2 * k2, q2 * k3...
  q3 * k1, q3 * k2, q3 * k3...]
The value matrix might be
[v1,v2,v3...]  
Note the q_i,k_i and v_i are vectors here and * denotes vector dot product.

So the query matmul key matrix matmul the value matrix will give
[q1 * k1 * v1 + q1*k2 *v1 + ....,
 q2 * k1 * v2 + q2*k2*v2 + ....,
 q3 * k1 * v3 ...q3*k2*v3 + ....]

Therefore, by applying a triangular mask like so:
[ q1 * k1, 0, 0...
  q2 * k1, q2 * k2, 0 ...
  q3 * k1, q3 * k2, q3 * k3...]

The query matmul key matmul value matrix, which gives the final weighted attention, will not have access to any of the
previous  q_i or k_i information.

**Comment by IpsumDominum** [comment]
I tried zeroing the attention, but I found that there is a difference. See image below. Left is zeroing and right is not zeroing. Clearly the right is learning and left is not.

Also with regards to ability to predict the entire sentence, and inference without teacher forcing, I think it is possible if you train the model enough. Have a try on the toy dataset, as expected, the model first learns how to predict the entire sequence well, given teacher training, but it takes a while for it to be able to learn to predict on its own predictions. Some training schemes might mitigate this like having semi teacher forcing, or deliberately skewed tokens in training. But I haven't tried it extensively.

![s](https://user-images.githubusercontent.com/30939978/152895446-7d051741-ccfb-4806-ae30-eda3b6a7f18c.png)

**Comment by IpsumDominum** [comment]
There is actually another problem which you made me realize. Which is the way I had done the indexing for the sequences, for instance, with the encoder:
```python3
        ###################
        # Encode English Sentence
        model.encode(item["english"][:, 1:-1])
        ###################
```
Here it should actually be just
```python3
        ###################
        # Encode English Sentence
        model.encode(item["english"])
        ###################
```
I did the indexing because initially when I constructed the dataset I appended a start token to all the sequences including the english ones, upon dataset generation (from .txt -> .npy  See file Extra_Scripts/gen_dataset.py). But I've changed that so this is no longer right.

The dataloader code (Dataset/translation_dataset.py)  isn't very clear and clean... I'll change it some time. But please take a look it might clear some questions.
For instance, I'm not sure if you were aware that I appended a "start" token to each of the german sentences already, inside of the dataloader code, so in your code you would be appending two "start" tokens.

```python3
g = item["german"].shape
x = torch.zeros( [g[0],g[1],],dtype=torch.long ).to(device)
all_outs = torch.tensor([],requires_grad=True).to(device)
for i in range(item["german"].shape[1]-1):
    xx = torch.zeros( [g[0],g[1], ],dtype=torch.long ).to(device)
    out = model(x)
    xx[:,i:i+1] = item["german"][:,i:i+1]
    x = x+xx
    all_outs = torch.cat((all_outs,out),dim=1)
```

**Comment by IpsumDominum** [comment]
Try the latest changes. The attention zeroing not effecting performance could be result of previous bugs. I can't remember if I tried it after I fixed the token things or not. :sweat_smile:

**Comment by shouldsee** [comment]
@IpsumDominum Sorry if I was not making myself clear. My point is basically, we cannot include the target token in the input sequence....

```python
print(item['german'][0].shape)
print(item['logits'][0].shape)
log = (item['logits'][0])
print([dataset.german_vocab_reversed[i] for i in item['german'][0]])
print([dataset.german_vocab_reversed[i] for i in item['logits'][0].argmax(axis=1) ])
```

```
torch.Size([31])
torch.Size([31, 4790])
['<start>', '', '(', 'Beifall', ' ', 'der', ' ', 'PSE', '-', 'Fraktion', ')', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>']
['<start>', '', '(', 'Beifall', ' ', 'der', ' ', 'PSE', '-', 'Fraktion', ')', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>', '<end>']
``` 

Using your current loss function as an example, on line A, you push the first (i) tokens of the target sentence into the model, and get a token prediction out, say i-th prediction. You then compare ith pred with (i+1)-th logit on line B. I think the new code should work now. Not sure what happened back then when I was running the old code....

input: [1,..,i] tokens
pred:  ith token
expect: [i+1] token

Bascially say i=3, you will be inputting ['<start>', '', '(', ], get a prediction ['blah'], and expect ['blah'] = ['Beifall'].

```python
        ###################
        # Output German, One Token At A Time
        all_outs = torch.tensor([], requires_grad=True).to(device)
        for i in range(item["german"].shape[1]-1):  
            #Minus one because we don't need to predict from <end> token
            out = model(item["german"][:, : i + 1])     ### line A
            all_outs = torch.cat((all_outs, out), dim=1)
        ###################
        
        ###################
        # Mask Out Extra Padded Tokens In The End(Optional)
        all_outs = all_outs * item["logit_mask"][:, 1:, :]                                               
        item["logits"] = item["logits"] * item["logit_mask"]  
        ###################

       loss = criterion(all_outs, item["logits"][:, 1:, :])   ### line B
```

**Comment by shouldsee** [comment]
BTW, I suggest using batchmean for the lossfunc, otherwise it's too small and has too many zeros to count....
```python
criterion = nn.KLDivLoss(reduction='batchmean')
```

**Comment by IpsumDominum** [comment]
Haha yes indeed having the output token in the input isn't right, and predicting the wrong token isn't right either :).
Good point on the reduction='batchmean', pytorch has been bugging with the warning message for a while. I'll change to that in new commit.

**Comment by IpsumDominum** [comment]
Thanks for finding the mistake!
